src/mscorlib/src/System/Lazy.cs

```diff
@@ -202,29 +239,171 @@ public Lazy(Func<T> valueFactory, bool isThreadSafe)
+        /// first object to complete its factory, uses that objects LazyWorker&lt;T&gt; implementation
+        /// </summary>
+        /// <param name="factory">The object factory used to create the underlying object</param>
+        /// <param name="comparand">The publication object, used for synchronisation and comparison</param>
+        /// <returns>The underlying object</returns>
+        internal T PopulateValuePublicationOnly(LazyImplementation initializer, LazyImplementation spinner, bool useDefaultConstructor)
+        {
+            T possibleValue;
+            if (useDefaultConstructor)
+            {
+                possibleValue = CreateViaDefaultConstructor();
+            }
+            else
+            {
+                var factory = m_factory;
+                if (factory == null)
+                    return spinner.GetValue(this);
+                possibleValue = factory();
+            }
+
+            try { }
+            finally
+            {
+                // we run this in a finally block to ensure that we don't get a partial completion due
+                // to a Thread.Abort, which could mean that other threads might be left in infinite loops
+                var previous = Interlocked.CompareExchange(ref m_implementation, spinner, initializer);
+                if (previous == initializer)
+                {
+                    m_value = possibleValue;
+                    m_factory = null;
+                    m_implementation = null;
+                }
+            }
+
+            return Value;
+        }
+
+        internal void WaitForPublicationOnly()
+        {
+            while (!ReferenceEquals(m_implementation, null))
+            {
+                // CreateValuePublicationOnly temporarily sets m_implementation to "this". The Lazy implementation
+                // has an explicit iplementation of LazyWorker which just waits for the m_value to be set, which is
+                // signalled by m_implemenation then being set to null.
+                Thread.Sleep(0);
+            }
+        }
+
+        #region Serialization
+        // to remain compatible with previous version, custom serialization has been added
+        // which should be binary compatible. Only valid values were ever serialized. Exceptions
+        // were thrown from the serializer, which halted it, if the Lazy object through an exception.
+
+        /// <summary>
+        /// wrapper class to box the initialized value, this is mainly created to avoid boxing/unboxing the value each time the value is called in case T is 
+        /// a value type
+        /// </summary>
+        [Serializable]
+        class Boxed
+        {
+            internal Boxed(T value)
+            {
+                m_value = value;
+            }
+            internal T m_value;
+        }
+
+        public Lazy(SerializationInfo information, StreamingContext context)
```

Comment:
I do not see test coverage for `Lazy<T>` serialization in corefx. It should be added to make sure that this works.